File: 지은/Lv.2/카펫.py
def solution(brown, yellow):
    brown -= 4
    brown //= 2

    for i in range(1, brown + 1):
        a, b = i, brown - i
        if a * b == yellow:
            return [max(a, b) + 2, min(a, b) + 2]
        

# 다른 풀이
def solution(brown, yellow):
    for h in range(1, int(yellow**0.5) + 1):
        if yellow % h == 0:
            w = yellow // h

            if 2 * (w + h) + 4 == brown:
                return [w + 2, h + 2]

File: 성재/Lv.2/카펫.java
/*
문제 링크 : https://school.programmers.co.kr/learn/courses/30/lessons/42842
 */

import java.util.*;

class Solution {
    public int[] solution(int brown, int yellow) {
        int[] answer = new int[2];
        int total_carpet = brown + yellow;

        for (int width = 3; width <= Math.sqrt(total_carpet); width++) {
            if ((total_carpet % width == 0) && (answer[1] < width)) {
                if ((width - 2) * (total_carpet / width - 2) == yellow) {
                    answer[1] = width;
                    answer[0] = total_carpet / answer[1];
                }
            }
        }

        return answer;
    }
}

/*
[약수 계산]
1. N의 약수를 구할 때는, 1부터 N의 제곱근 까지의 수만 0으로 나누어 떨어지는지 확인
2. N을 1에서 구한 수로 나누게 되면 모든 약수들을 구할 수 있다!
 */
